### PR TITLE
Add label support

### PR DESCRIPTION
**Pull request by mlouielu**
Label syntax as `<LABEL>:` in one separate line, it can not do as `<LABEL>: add $0 $0 #1`, jump instructions can jump to label with `:<LABEL>` syntax (e.g. `jmp :label`, `jnz #1 :label`).

`<LABEL>` is valid in any alphabet and digit, it should be fixed.

### CODE REVIEW

File: as.c [hunk 1]
```diff
@@ -38,13 +39,25 @@ static inline vm_operand make_operand(vm_env *env, char *line, const char *data)
...
         break;
+    case ':':
+        /*
+         * This will not fill the label's next pc to op.value.id
+         * when calling `make_operand`.
+         * When `vm_add_inst`, it will add this label reference into env, then
+         * call `vm_hook_label` to get the actuall label's next_pc
+         */
+        op.type = LABEL;
+        op.value.id = -1;  /* Label not found */
```
Comment (jserv) [inline as.c]: Don't hardcode `-1`. Use `enum` instead.

File: as.c [hunk 2]
```diff
@@ -84,6 +97,13 @@ static void assemble_line(vm_env *env, char *line)
     const struct instruction *inst = find_inst(mnemonic);
 
     if (!inst) {
+        /* Check label */
+        if (strlen(mnemonic) > 1 &&
```
Comment (jserv) [inline as.c]: It is weird to check `strlen(mnemonic) > 1` for C-style string.

File: as.c [hunk 3]
```diff
@@ -84,6 +97,13 @@ static void assemble_line(vm_env *env, char *line)
...
+                mnemonic[strlen(mnemonic) - 1] == ':') {
```
Comment (jserv) [inline as.c]: `strlen(mnemonic)` is not always changed here. You don't have to invoke it several times.

File: vm.h [hunk 4]
```diff
@@ -31,16 +32,52 @@ typedef struct {
...
+} vm_env;
 
 vm_env *vm_new();
 void vm_free(vm_env *);
 
 size_t vm_add_const(vm_env *, int, void *);
 size_t vm_add_inst(vm_env *, vm_inst);
+size_t vm_add_label(vm_env *, const int, char *);
+int vm_find_label(vm_env *, const char *);
+void vm_hook_label(vm_env *);
```
Comment (jserv) [inline vm.h]: `vm_hook_label` is not intuitive. Using `register` or `insert` might be a better alternative.
Terms: http://arantxa.ii.uam.es/~modonnel/Compilers/04_SymbolTablesI.pdf

File: vm.h [hunk 5]
```diff
@@ -31,16 +32,52 @@ typedef struct {
...
+    char *label;
+    int next_pc;
+} vm_label;
+
 #define VM_T(_op) _op->type
 #define VM_INT(_op) _op->value.vint
 
-typedef struct __vm_env vm_env;
+/* Constant pool max size */
+#define CPOOL_MAX_SIZE 100
```
Comment (jserv) [inline vm.h]: Don't expose implementation details in public header(s).

File: as.c (hunk 2 above)
Comment (mlouielu) [inline as.c]: This check want to avoid label only contain `:`, so it will need to check the length is greater than 1.

File: tests/label.s [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+go:
+add #0 #1 #0    ; result = result + n
+sub #1 $1 #1    ; n = n - 1
+jnz #1 :go      ; jump to go if #1 not zero
+
+jmp :output       ; jump to output
+add $0 $100 #1    ; should not be executed, #1 remained 0
+; print the result
+output:
+print ^1+2+...+10=
```
Comment (jserv) [inline tests/label.s]: Hack `^` was removed, and you should use `print "1 + 2 + 3 + ... + 10"`.

File: vm.c [hunk 7]
```diff
@@ -108,10 +124,45 @@ size_t vm_add_const(vm_env *env, int type, void *value)
...
+    env->labels[env->labels_count].next_pc = env->insts_count;
+
+    return env->labels_count++;
+}
+
+int vm_find_label(vm_env *env, const char *label)
+{
+    /* FIXME: Better label search method */
+    for (int i = 0; i < env->labels_count; ++i) {
+        if (strcmp(env->labels[i].label, label) == 0) {
```
Comment (jserv) [inline vm.c]: To simplify, rewrite it as `if (!strcmp(env->labels[i].label, label)) {`

File: vm.h [hunk 8]
```diff
@@ -31,6 +32,14 @@ typedef struct {
     vm_handler handler;
 } vm_opcode_impl;
 
+typedef struct {
+    enum { SYMBOLIC, NUMERIC } type;
+    char *label;
+    int next_pc;
+} vm_label;
+
+#define LABLE_NOT_FOUND -1
```
Comment (jserv) [inline vm.h]: Using `typedef enum { LABLE_NOT_FOUND = -1, ERR1, ERR2, ERR3, ... } vm_err_t;` is better for diagnosis.

File: tests/label.s [hunk 9]
```diff
@@ -0,0 +1,16 @@
+; Test for label
+; Output 1 + 2 + ... + 10 result
+add $0 $0 #0    ; result
+add $0 $10 #1   ; start from 10
+go:
+add #0 #1 #0    ; result = result + n
+sub #1 $1 #1    ; n = n - 1
+jnz #1 :go      ; jump to go if #1 not zero
```
Comment (jserv) [inline tests/label.s]: The notation `:go` looks a bit confusing. Can you accept the natural form `go`?

File: vm.h [hunk 10]
```diff
@@ -31,6 +32,19 @@ typedef struct {
...
 
+typedef struct {
+    enum { SYMBOLIC, NUMERIC } type;
+    char *label;
+    int next_pc;
+} vm_label;
+
+typedef enum {
+    LABEL_NOT_FOUND = -1,
+    UNIMPLEMNET_ERR1,
```
Comment (jserv) [inline vm.h]: Leaving one `UNIMPLMENTED` should be enough. You can rewrite parts of file `opcode.c` manipulations.

File: vm.h [hunk 11]
```diff
@@ -40,9 +40,7 @@ typedef struct {
 
 typedef enum {
     LABEL_NOT_FOUND = -1,
-    UNIMPLEMNET_ERR1,
-    UNIMPLEMNET_ERR2,
-    UNIMPLEMNET_ERR3,
+    UNIMPLEMNET,
```
Comment (jserv) [inline vm.h]: Should be "UNIMPLEMENTED".

File: tests/label.expected [hunk 12]
```diff
@@ -0,0 +1,3 @@
+10! =
```
Comment (jserv) [inline tests/label.expected]: Labels should exist in `coverage.s` as well.

File: as.c [hunk 13]
```diff
@@ -189,17 +190,15 @@ static inline vm_operand make_operand(vm_env *env, char *line, const char *data)
         op.value.id = vm_add_const(env, STR, quoted_strdup(data));
         break;
     default:
-        printf(
```
Comment (jserv) [inline as.c]: Why did you delete these?

File: vm.c [hunk 14]
```diff
@@ -25,14 +26,22 @@
 
 #define END_OPCODES
 
-#define VM_GOTO(n) \
-    pc = n;        \
+#define VM_GOTO(n)  \
```
Comment (jserv) [inline vm.c]: Use `clang-format` to indent this file.

File: vm.c [hunk 15]
```diff
@@ -25,14 +26,22 @@
...
+    if (OPCODE.op1.type == LABEL && n == -1) {  \
+        printf("Error: Undefined label \"%s\".\n", OPCODE.op1.label); \
```
Comment (jserv) [inline vm.c]: Since macro `FATAL` is introduced in header `private.h`, you can rewrite with it.

File: vm.c [hunk 16]
```diff
@@ -25,14 +26,22 @@
...
+    }                                         \
+    pc = n;                                   \
     goto *labels[OPCODE.opcode]
 
 #define VM_J_TYPE_INST(cond)                                     \
     do {                                                         \
         int gle = vm_get_op_value(env, &OPCODE.op1)->value.vint; \
         if (gle cond 0) {                                        \
+            if (OPCODE.op2.type == LABEL && OPCODE.op2.value.id == -1) {      \
+                printf("Error: Undefined label \"%s\".\n", OPCODE.op2.label); \
```
Comment (jserv) [inline vm.c]: Ditto. Use macro `FATAL`.

File: vm.c [hunk 17]
```diff
@@ -68,14 +77,21 @@
...
+    vm_value cpool[CPOOL_MAX_SIZE];               /* Constant pool */
+    vm_value temps[TEMPS_MAX_SIZE];               /* Temporary storage */
+    vm_opcode_impl impl[OPCODE_IMPL_MAX_SIZE];    /* OPCODE impl */
+    vm_label labels[LABELS_MAX_SIZE];             /* Labels */
+    vm_operand *label_reference[LABELS_MAX_SIZE]; /* Operand's label reference */
     int insts_count;
     int cpool_count;
     int temps_count;
+    int labels_count;
+    int label_reference_count;
```
Comment (jserv) [inline vm.c]: `_reference_count` is too long. Use `_refcnt` instead.

File: vm.c [hunk 18]
```diff
@@ -115,10 +131,45 @@ size_t vm_add_const(vm_env *env, int type, void *value)
...
+{
+    /* FIXME: Better label search method */
+    for (int i = 0; i < env->labels_count; ++i) {
+        if (!strcmp(env->labels[i].label, label)) {
+            return env->labels[i].next_pc;
+        }
+    }
+
+    /* XXX: Need something like `PyErr_SetString` for error handling */
+    printf("Label Register Error: Label \"%s\" not found\n", label);
```
Comment (jserv) [inline vm.c]: Ditto. Use macro `FATALX`.

**Comment by jserv** [comment]
@mlouielu, Since `call` and `ret` instructions are implemented by @pakls, the label support is getting important for us to write more non-trivial applications. Would you like to revise these above changes?

**Comment by mlouielu** [comment]
@jserv Sure, I will get another round to revising this. (I'm on the vacation these days.)

**Comment by jserv** [comment]
@mlouielu, would you like to continue this work recently? If so, I expect to implement function call support.

**Comment by mlouielu** [comment]
@jserv I think this should propose by a new round of PR. This PR has too much conflict to current code.

**Comment by jserv** [comment]
Close per the suggestion from @mlouielu
